qiskit/transpiler/_passmanager.py

```diff
@@ -1,33 +1,284 @@
-    do this correctly & efficiently
-    (i.e. keep track of dependencies between passes)
+    def add_control_flow_plugin(self, name, control_flow_plugin):
+        """
+        Adds a control flow plugin.
+        Args:
+            name (string): Name of the plugin.
+            control_flow_plugin (ControlFlowPlugin): The class implementing a control flow plugin.
+        """
+        self.working_list.add_control_flow_plugin(name, control_flow_plugin)
+
+    def remove_control_flow_plugin(self, name):
+        """
+        Removes a control flow plugin.
+        Args:
+            name:
+        """
+        self.working_list.remove_control_flow_plugin(name)
+
+
+class WorkingList():
+    """
+    A working list is the way that a pass manager organizes the
+    schedule of things to do.
     """
+
     def __init__(self):
-        """Initialize an empty PassManager object
-        (with no passes scheduled).
+        self.list_of_items = []
+        self.control_flow_plugins = {'condition': PluginConditional,
+                                     'do_while': PluginDoWhile}
+
+    def add_control_flow_plugin(self, name, control_flow_plugin):
+        """
+        Adds a control flow plugin.
+        Args:
+            name (string): Name of the plugin.
+            control_flow_plugin (ControlFlowPlugin): The class implementing a control flow plugin.
+        """
+        self.control_flow_plugins[name] = control_flow_plugin
+
+    def remove_control_flow_plugin(self, name):
+        """
+        Removes the plugin called name.
+
+        Args:
+            name (string): The control flow plugin to remove.
+        Raises:
+            KeyError: If the name is not found.
         """
-        self._passes = []
+        if name not in self.control_flow_plugins:
+            raise KeyError("Control flow plugin not found: %s" % name)
+        del self.control_flow_plugins[name]
+
+    def add(self, passes, **control_flow_plugins):
+        """
+        Populates the working list with passes.
+
+        Args:
+            passes (list): a list of passes to add to the working list.
+            control_flow_plugins (kwargs): See add_control_flow_plugin(). Dictionary of control flow
+                plugins. Defaults:
+                do_while (callable property_set -> boolean): The pass (or passes) repeats until the
+                   callable returns False.
+                   Default: lambda x: False # i.e. passes run once
+                condition (callable property_set -> boolean): The pass (or passes) runs only if the
+                   callable returns True.
+                   Default: lambda x: True # i.e. passes run
+        """
+        for control_flow, condition in control_flow_plugins.items():
+            if condition and control_flow in self.control_flow_plugins:
+                self.list_of_items.append(
+                    self.control_flow_plugins[control_flow](passes, **control_flow_plugins))
+                break
```

Comment:
Are you breaking because you only allow one flow controller per call to `add`? If this is so, how do you ensure when using the `condition` **and**  `do_while` parameters that you will register a `PluginCondition` instance and not a `PluginDoWhile` one?

Reply:
Not anymore. Also part of https://github.com/Qiskit/qiskit-terra/commit/673774adae838b4bc017eaa36897d27c96597c41